<font size="8">Homework 2: Mini Project - AirBnb Document Database</font>

Group number: `24`

Group members:
1. `Ana Miguel Monteiro : 20221645`
2. `Ana Rita Viseu : 20220703`
3. `Ana Sofia Vieira Mendonça : 20220678`
4. `Sara Galguinho : 20220682`
5. `Susana Teresa Dias : 20220198`

The Homework 2 is comprised of two parts:
1. Data modelling
2. Queries to database to answer the questions

<font size="6">1. Data Modelling</font>

<font size="4">Congratulations! You have been successfully hired as the new Data Engineering and Management team in the AirBNB Business Intelligence department. 
The organization is undergoing restructuring as the performance of the previous teams was not believed to be satisfactory.

Before leaving their post, the head of the Data Modelling department left this note of identified issues.

Currently a lot of data related to AirBNB listings is all collected in the same document. While this has some advantages it has also caused issues. The query performance is known to be too slow, so some decisions on hybridization of data models are required. The previous team did not have experience in the application of patterns, which has been identified as a key area to improve performance through the enhancement of the data model. Indexes have also not been previously applied.

The AirBNB application use has been growing rapidly, especially the the number of reviews. Because of that we have been overwriting them regularly, but the business intelligence department will benefit from storing all the Reviews and analysing them over time. 

Finally, there are some mistakes in the data collection : for instance, writing the same data multiple times, and assigning wrong timestamps for when transactions occurred. The new team will need to decide how to address these.   

In your new role, you are expected to think about the database use case in each question, analyse all the information you have, how commonly is this query needed, how heavy will it be on reads and write etc. You should modify the database schema, where you deem appropriate, to optimise for the business use cases. To do so you should use all the tools you have learned in the lab and lectures, especially embedding and linking, indexes, and patterns. This can involve creating new fields, documents, and collections. Indicate which pattern you are applying and your reasoning for doing so in comments, remember to identify any duplication that is required and any risks of staleness. 

Remember to think about the relevancy of all data that will be returned in each case and if this is needed or adding excessive time to return:
1. to streamline the data collection system;
2. to clean up the data and optimise what will be returned for each use case;
3. to use correct patterns to improve speed of typical queries;
4. to ensure that all departments receive correct and relevant information from the database;
5. to provide the data model schema to inform other departments of the changes.

Good practices in our company include:
1. all newly created fields have CAPITALIZED names;
2. all new queries function on the most updated version of the database (if you make multiple changes all your queries still work after the final changes);
3. for some of the queries you will likely want to change the database schema in some way. Each database transformation choice is properly documented using format:
    * `TRANSFORMATION APPLIED: <NAME and DESCRIPTION OF TRANSFORMATION>`
    * `REASONING BEHIND: <WHY YOU DID IT>`
    * `EXPECTED RESULT: <WHAT IS THE (EXPECTED) RESULT OF RUNNING QUERIES GIVEN THIS TRANSFORMATION>`

</font>


**Initial data setup:**

Before starting on the queries below review the data and look to adjust the schema for the typical use case described below.

1) The most typical use case for the database is to show information relating to a property listing to a customer. This is done by a query to the database which returns one of the listing documents. Currently a lot of time is spent when a listing is retrieved from the database to show to a customer. Decide what information should be returned in a typical query and optimise the structure for this use case. For example, we typically only want a sample of reviews but not all reviews (although all reviews ), the customer does not need to know past transaction data, etc. Update the document schema for this typical use case. This may involve the creation of new collections and documents.  

2) Review the data for any errors (such as transactions that don’t fit the listing) or inappropriate duplication and clean up as appropriate. 


<font size="6">2. Uses for the database</font>

<font size="4">Different AirBNB departments require different analytics from our common database.
Here are the list of all specific questions from different departments: 

**Standard Dificulty Questions:**

3)	Once per month we like to reward hosts with recognition. Pick three superhosts who have at least two property listings that can accommodate more than four people?

4)	One of employees is thinking of buying a property to rent out. Which bed type is most common in the listings that have waterfront and a dishwasher in New York?

5)	We are thinking of identifying someone to hire to write review for us professionally. What is the name of the reviewer who left the longest review in New York?

6)	We are thinking of trying to assess the security of different areas based on the security deposit required. What is the biggest and smallest difference between the price and deposit for security per number of visitors staying at the property?

7)	We want to identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. We will use this information to target advertising of different customers. It is not expected this information will change much over time so we won’t look to update it we just would like a current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Advanced Dificulty Questions (Remember to consider how to optimise the database for these queries):**

8)	We want to have a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

9)	We want to start tracking our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).


10)	For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We are thinking we may soon add more metrics, although we don’t know what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

11)	We wish to be able to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.


12)	We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

**Database updates:**

After you have completed the above tasks and have optimised your database show how you would complete each of the updates below. You can create whatever fictional data you like for the below. Remeber to cosider how to prevent any data you have previously created from becoming stale.

13)	Add a new property, this property should have a new host, and should be located in one of the top 10 cities. The host selects the top 10 most common amenities to be listed for the property.

14)	Add a new review to this property, the review should be from one of our top 20 reviewers.

15)	Add a new review metric called x_factor to the new document and give a score of 10. Show that the average across all metrics is correctly calculated for this listing, with the query you previously developed.


# Setup

1. run Jupyter 
2. run Docker
3. create MongoDB container in Command Prompt: make sure you don't have 'leftover' container from previous runs

        docker run --name mongodb -d -e MONGO_INITDB_ROOT_USERNAME=AzureDiamond -e MONGO_INITDB_ROOT_PASSWORD=hunter2 -p 27017:27017 mongo

4. run Studio3T

5. create New Connection OR reconnect to existing Connection:

    5.A	In Studio3T create Connection:
            a.	Press Connection -> New Connection
            b.	Insert credentials into URI field: mongodb://AzureDiamond:hunter2@localhost:27017
            c.	Press Test Connection
            d.	Assign a name to the Connection (top empty line) -> Save
            e.	Press Connect
            
    5.B In Studio3T reconnect to Connection:
            a. press Connect (top left corner)
            b. choose MongoDB connection  -> press Connect (MongoDB container must be runnning by this time)
            
6.	In Studio3T import the Database:
            a.	Press Import
            b.	Choose BSON mongodump archive 
            c.	Find your database path
            d.	Select All file formats and choose the database file ‘sampledata.archive’
            e.	Press Run: be patient while the database loads
            f.	Inspect the collections and documents

Done!

In [1]:
# Python Connector

#!pip install pymongo
# or #!conda install -y pymongo

from datetime import datetime
from pprint import pprint
import time
from bson.objectid import ObjectId

from pymongo import MongoClient

user="AzureDiamond"
password="hunter2"
host="localhost"
port="27017"
protocol="mongodb"

client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

# Database check
db = client.sample_airbnb 
print(f"Database info: {db}\n")
db.name 

Database info: Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'sample_airbnb')



'sample_airbnb'

# Loading and Inspecting database


In [2]:
# Collections are inside our Database 'sample_analytics'

collection_list = db.list_collection_names()

print(f"The database contains {len(collection_list)} collections")
print(f"All collections: {collection_list[0:]}")
print(f"Collection {collection_list[0]} contains {db[collection_list[0]].count_documents({})} documents")

The database contains 1 collections
All collections: ['listingsAndReviews_new']
Collection listingsAndReviews_new contains 5555 documents


In [3]:
#db.listing_customer.drop()
#db.listing_address.drop()
#db.listing_reviews.drop()
#db.listing_amenities.drop()

[PyMongo documentation](https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html)

### Question 1

In [4]:
# Find one document in the collection 'listingsAndReviews_new'
listing_1 = db.listingsAndReviews_new.find_one() # Assign and import the database

# Check the fields and the document itself
pprint(listing_1)

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

In [5]:
# Specifying the "listingsAndReviews_new" collection as a source
db.listingsAndReviews_new.aggregate([
    # Using the $project to specify the fields returned
    {"$project": {"id": 1, 
                "access":1, 
                "accommodates":1,
                "amenities":1,
               "bathrooms":1,
               "bed_type":1,
               "bedrooms":1,
               "beds":1,
               "cancellation_policy":1,
               "cleaning_fee":1,
               "description":1,
               "extra_people":1,
               "guests_included":1,
               "host_about":1,
               "host_id":1,
               "host_identity_verified":1,
               "host_is_superhost":1,
               "host_listings_count":1,
               "host_location":1,
               "host_name":1,
               "host_picture_url":1,
               "host_response_rate":1,
               "host_response_time":1,
               "host_url":1,
               "house_rules":1,
               "images":1,
               "interaction":1,
               "last_review":1,
               "listing_url":1,
               "maximum_nights":1,
               "minimum_nights":1,
               "name":1,
               "neighborhood_overview":1,
               "notes":1,
               "price":1,
               "property_type":1,
               "review_scores_rating":1,
               "room_type":1,
               "security_deposit":1,
               "space":1,
               "summary":1,
               "transit":1,
               "address":1}},
    {
    # Using the $out operator to create a new collection named "listing_customer"
    '$out': "listing_customer"
    }
])

Let's create the ``listing_reviews`` collection. We'll make use of the ``referential linking`` to create this new collection, since there are embedded documents inside the reviews field.

In [6]:
db.listingsAndReviews_new.aggregate([
    # Using the unwind function to separate each field into a new document
  { "$unwind": "$reviews" },
  { "$project": { 
      "_id": "$reviews._id",
      "listing_id": "$reviews.listing_id",
      "reviewer_id": "$reviews.reviewer_id",
      "reviewer_name": "$reviews.reviewer_name",
      "comments": "$reviews.comments",
      "date": "$reviews.date"
  }},
    # Using the $out to create a new collection named "listing_reviews"
  { "$out": "listing_reviews" }
])

### Question 2 - Review the data for any errors (such as transactions that don’t fit the listing) or inappropriate duplication and clean up as appropriate.


We'll look for imputation errors or incoherencies that might exist in the databse.

- 1. Look for records where "beds", "bathrooms" or "price" is 0. It might not be the best approach to keep these kinds of records.
- 2. Value Counts or other incoherencies
    - 2.1. "maximum_nights" field
    - 2.2. "minimum_nights" field
    - 2.3. Duplicates  in the "id" field
    - 2.4. "summary" field
    - 2.5. "description" field
    - 2.6. "notes" field
    - 2.7. "access" field
    - 2.8. "amenities" field
    - 2.9. "accommodates" field
    - 2.10. "bathrooms" field
    - 2.11. "bedrooms" field
    - 2.12. "bed" field
    - 2.13. "extra_people" field
    - 2.14. "guests_included" field
    - 2.15. "host_identity_verified" field
    - 2.16. "host_is_superhost" field
    - 2.17. "host_listings_count" field
    - 2.18. "host_name" field
    - 2.19. "price" field
    - 2.20 "property_type" field
    - 2.21 "room_type" field

- 3. It might not be the best approach to keep records where the number of guests included is higher than the number of accomodates an Airbnb accomodates.

#### **1. Record where "beds", "bathrooms" or "price" is zero.**

In [7]:
# Filter the documents where the value of the "beds", "bathrooms", or "price" field is 0

cursor = db.listing_customer.aggregate([
  {"$match": {"$or": [{"beds": 0}, {"bathrooms": 0}, {"price": 0}]}},
    
# Grouping the selected documents into a single group represented by the _id value of 0. 
# The $sum operator is used to count the number of documents in this group that match the condition in 
# the $match stage. The result is returned in the "count" field.
    
  {"$group": {"_id": 0, "count": {"$sum": 1}}}
])

results = list(cursor)
print(results)

[{'_id': 0, 'count': 43}]


In [8]:
# Deleting the records where the value of the "beds", "bathrooms", or "price" field is 0

db.listing_customer.delete_many({"$or": [{"beds": 0}, {"bathrooms": 0}, {"price": 0}]})

#### 2. Value Counts 

##### 2.1. "maximum_nights" field

In [9]:
# Counting the number of times each unique value of maximum_nights appears
value_counts = db.listing_customer.aggregate([
    # Converting the maximum_nights field to integer using toInt()
    {"$project": {"maximum_nights_int": {"$toInt": "$maximum_nights"}}},
    # Grouping by the field maximum_nights_int and counting the number of times each value appears in the collection
    {"$group": {"_id": "$maximum_nights_int", "count": {"$sum": 1}}},
    # Sorting the results by descending order of the "maximum_nights_int" field
    {"$sort": {"_id": -1}},
    {"$project": {"_id":0, "Maximum Nights": "$_id", "Count": "$count"}}
])


# Printing the output using pprint
pprint(list(value_counts))

[{'Count': 3, 'Maximum Nights': 2147483647},
 {'Count': 1, 'Maximum Nights': 1234567890},
 {'Count': 1, 'Maximum Nights': 333333},
 {'Count': 3, 'Maximum Nights': 10000},
 {'Count': 1, 'Maximum Nights': 5000},
 {'Count': 1, 'Maximum Nights': 2800},
 {'Count': 1, 'Maximum Nights': 1825},
 {'Count': 1, 'Maximum Nights': 1250},
 {'Count': 1, 'Maximum Nights': 1126},
 {'Count': 3137, 'Maximum Nights': 1125},
 {'Count': 93, 'Maximum Nights': 1124},
 {'Count': 7, 'Maximum Nights': 1123},
 {'Count': 1, 'Maximum Nights': 1121},
 {'Count': 13, 'Maximum Nights': 1120},
 {'Count': 1, 'Maximum Nights': 1117},
 {'Count': 2, 'Maximum Nights': 1111},
 {'Count': 1, 'Maximum Nights': 1110},
 {'Count': 1, 'Maximum Nights': 1109},
 {'Count': 4, 'Maximum Nights': 1100},
 {'Count': 3, 'Maximum Nights': 1095},
 {'Count': 1, 'Maximum Nights': 1086},
 {'Count': 1, 'Maximum Nights': 1084},
 {'Count': 1, 'Maximum Nights': 1003},
 {'Count': 1, 'Maximum Nights': 1001},
 {'Count': 15, 'Maximum Nights': 1000},
 {'C

Our team noticed some values that are strange, for instance, ``Maximum Nights`` higher than 3 years, or the values "333333", "2147483647" or "1234567890".

Additionally, we also noticed that the most common ``Maximum Nights`` is 1125, with 3137 Airbnbs with that number in the maximum_nights field. Because of that, our team decided to delete every record with ``Maximum Nights`` higher than 1125. 

To perform this, we will take into consideration that the ``Maximum Nights`` field is stored as a string, therefore we will use the ``parseInt`` funtion that converts a string to an integer. Then, we remove all fields with a higher value than 1125.

In [10]:
result = db.listing_customer.find({
    # Convert maximum_nights to an integer, and filter the documents where Maximum Nights is greater than 1125
    "$where": "parseInt(this.maximum_nights) > 1125"
}, {"_id": 0, "maximum_nights": 1})

# Using the delete_many function, we delete the documents where Maximum Nights is greater than 1125
result = db.listing_customer.delete_many({
    "$where": "parseInt(this.maximum_nights) > 1125"
})

# Printing the number of documents that were deleted
print(f"{result.deleted_count} documents deleted.")

13 documents deleted.


##### 2.2. "minimum_nights" field

In [11]:
value_counts = db.listing_customer.aggregate([
    # Converting the minimum_nights field to integer using toInt()
    {"$project": {"minimum_nights_int": {"$toInt": "$minimum_nights"}}},
    # Grouping by the field minimum_nights_int and counting the number of times each value appears in the collection
    {"$group": {"_id": "$minimum_nights_int", "count": {"$sum": 1}}},
    # Sorting the results by descending order of the "minimum_nights_int" field
    {"$sort": {"_id": -1}},
    {"$project": {"Minimum Nights": "$_id", "Count": "$count", "_id":0}}
])

pprint(list(value_counts))

[{'Count': 2, 'Minimum Nights': 365},
 {'Count': 1, 'Minimum Nights': 300},
 {'Count': 1, 'Minimum Nights': 200},
 {'Count': 12, 'Minimum Nights': 180},
 {'Count': 1, 'Minimum Nights': 179},
 {'Count': 1, 'Minimum Nights': 160},
 {'Count': 4, 'Minimum Nights': 120},
 {'Count': 1, 'Minimum Nights': 115},
 {'Count': 1, 'Minimum Nights': 100},
 {'Count': 1, 'Minimum Nights': 96},
 {'Count': 12, 'Minimum Nights': 90},
 {'Count': 1, 'Minimum Nights': 80},
 {'Count': 15, 'Minimum Nights': 60},
 {'Count': 1, 'Minimum Nights': 50},
 {'Count': 2, 'Minimum Nights': 45},
 {'Count': 1, 'Minimum Nights': 40},
 {'Count': 1, 'Minimum Nights': 35},
 {'Count': 79, 'Minimum Nights': 32},
 {'Count': 46, 'Minimum Nights': 31},
 {'Count': 114, 'Minimum Nights': 30},
 {'Count': 25, 'Minimum Nights': 29},
 {'Count': 35, 'Minimum Nights': 28},
 {'Count': 3, 'Minimum Nights': 27},
 {'Count': 1, 'Minimum Nights': 26},
 {'Count': 4, 'Minimum Nights': 25},
 {'Count': 1, 'Minimum Nights': 22},
 {'Count': 12, 'Mini

Regarding these ``Minimum Nights`` values, we consider that there are no values to be removed, since there can be 1 minimum night and the highest values (like 365), are possible as there is long-term rental in Airbnb.

##### 2.3. Duplicates  in the "id" field

In [12]:
# Check for duplicates in the field "id"
duplicates_id = db.listing_customer.aggregate([
    # Grouping by the field "id" and counting how many times each id appears in the collection listing_customer
    {"$group": {"_id": "$_id", "Count": {"$sum": 1}}},
    # Show only if there are duplicates (greater than 1)
    {"$match": {"Count": {"$gt": 1}}}])

# Check the count 
results = list(duplicates_id)
print(results)

[]


##### 2.4. "summary" field

In [13]:
duplicates_summary = db.listing_customer.aggregate([
    # Find documents where the summary field does not exist or is null
    {"$match": {"$or": [{"summary": {"$exists": False}}, {"summary": {"$in": [".", " ", "  ", "?",""]}}]}},
    # Project the summary field
    {"$project": {"summary": 1}}
])

# Check the count
results = list(duplicates_summary)
pprint(results)

[{'_id': '1003530', 'summary': ''},
 {'_id': '10133554', 'summary': ''},
 {'_id': '10199760', 'summary': ''},
 {'_id': '1022200', 'summary': ''},
 {'_id': '1036027', 'summary': ''},
 {'_id': '1038163', 'summary': ''},
 {'_id': '1042446', 'summary': ''},
 {'_id': '1068505', 'summary': ''},
 {'_id': '1073607', 'summary': ''},
 {'_id': '1077686', 'summary': ''},
 {'_id': '1079887', 'summary': ''},
 {'_id': '1090153', 'summary': ''},
 {'_id': '1098211', 'summary': ''},
 {'_id': '1117706', 'summary': ''},
 {'_id': '1132409', 'summary': ''},
 {'_id': '1144090', 'summary': ''},
 {'_id': '1163119', 'summary': ''},
 {'_id': '1182475', 'summary': ''},
 {'_id': '1195977', 'summary': ''},
 {'_id': '1209353', 'summary': ''},
 {'_id': '1215226', 'summary': ''},
 {'_id': '121861', 'summary': ''},
 {'_id': '1227759', 'summary': ''},
 {'_id': '1229106', 'summary': ''},
 {'_id': '1290112', 'summary': ''},
 {'_id': '1294944', 'summary': ''},
 {'_id': '1310308', 'summary': ''},
 {'_id': '1332929', 'summar

There are some documents with an empty ``summary``. In fact, some Airbnb customers might find this field important, however, our team does not consider it essential, since one can understand an Airbnb by the remaining information. Therefore, in the next steps we will check if the ``summary`` **and** ``description`` fields are empty, and if so, delete those documents. We will do this in the next "chapter".

Now, we will check for duplicates in the ``summary`` field, since it is strange if two different Airbnbs to have the same summary.

In [14]:
# Check for Duplicates in the field summary
duplicates_summary = db.listing_customer.aggregate([
    # Grouping by the "summary" field and counting how many times each id appears in the collection listing_customer
    {"$group": {"_id": "$summary", "count": {"$sum": 1}}},
    # Show only if there are duplicates (greater than 1)
    {"$match": {"count": {"$gt": 1}}},
    # Sun the number of duplicates
    {"$group": {"_id": None, "Nr of duplicates": {"$sum": 1}}},
    # Project the "Nr of duplicates" field
    {"$project": {"_id": 0, "Nr of duplicates": 1}}
])

# Check the count 
results = list(duplicates_summary)
print(results)

[{'Nr of duplicates': 34}]


There are 34 duplicates in the ``summary`` field. However, it might not be a problem if the duplicates belong to the same host, who might have similar or even exact the same listings. For example, an apartment building where the host is listing the different appartments he has, or terraced houses that are all lookalike. Therefore, we decided not to delete these duplicates.

We will check for duplicated ``summary`` fields between different hosts, because this might not be an error, and therefore should be deleted.

In [15]:
duplicates_summary = db.listing_customer.aggregate([
    # Group by host_id and summary, and count the number of occurrences
    {"$group": {"_id": {"host_id": "$host_id", "summary": "$summary"}, "count": {"$sum": 1}}},
    # Group by summary only, and count the number of groups with more than one host_id
    {"$group": {"_id": "$_id.summary", "Nr of duplicates": {"$sum": {"$cond": [{"$gt": ["$count", 1]}, 1, 0]}}}},
    # Show only if there are duplicates (greater than 1)
    {"$match": {"Nr of duplicates": {"$gt": 1}}},
    # Project the number of duplicates
    {"$project": {"_id": 0, "Nr of duplicates": 1}}
])

# Check the count 
results = list(duplicates_summary)
pprint(results)

[{'Nr of duplicates': 6}]


There are 6 duplicated `summary` fileds among different hosts. Considering that this might be an error, we decided to delete these 6 documents.

In [16]:
# Find the duplicated documents
duplicates = db.listing_customer.aggregate([
    # Group by host_id and summary, and count the number of occurrences
    {"$group": {
        "_id": {"host_id": "$host_id", "summary": "$summary"},
        "ids": {"$addToSet": "$_id"},
        "count": {"$sum": 1}
    }},
    # Show only if there are duplicates (greater than 1)
    {"$match": {"count": {"$gt": 1}}}
])

# Delete the duplicated documents
for doc in duplicates:
    ids_to_delete = doc['ids'][1:]
    db.listing_customer.delete_many({"_id": {"$in": ids_to_delete}})

##### 2.5. "description" field

Let's check if the `summary` and `description` fields are both empty. If so, delete those records.

In [17]:
duplicates_both = db.listing_customer.aggregate([
    # Find documents where both the summary and description fields are empty or null
    {"$match": {"$and": [{"$or": [{"summary": {"$exists": False}}, {"summary": {"$in": [".", " ", "  ", "?",""]}}]}, 
                         {"$or": [{"description": {"$exists": False}}, {"description": {"$in": [".", " ", "  ", "?",""]}}]}]}},
    # Project the summary and description fields
    {"$project": {"summary": 1, "description": 1}}
])

# Check the count
results = list(duplicates_both)
pprint(results)

[{'_id': '10199760', 'description': '', 'summary': ''},
 {'_id': '15908244', 'description': '', 'summary': ''},
 {'_id': '15949042', 'description': '', 'summary': ''},
 {'_id': '19830476', 'description': '', 'summary': ''},
 {'_id': '19830549', 'description': '', 'summary': ''},
 {'_id': '19967176', 'description': '', 'summary': ''},
 {'_id': '21822223', 'description': '', 'summary': ''},
 {'_id': '24505019', 'description': '', 'summary': ''},
 {'_id': '24740074', 'description': '', 'summary': ''},
 {'_id': '25150521', 'description': '', 'summary': ''},
 {'_id': '25484442', 'description': '', 'summary': ''},
 {'_id': '25939348', 'description': '', 'summary': ''},
 {'_id': '26008844', 'description': '', 'summary': ''},
 {'_id': '26044607', 'description': '', 'summary': ''},
 {'_id': '26060837', 'description': '', 'summary': ''},
 {'_id': '26064830', 'description': '', 'summary': ''},
 {'_id': '26242605', 'description': '', 'summary': ''},
 {'_id': '26273293', 'description': '', 'summary

Since the majority of the customers might find having at least one of these two fields filled when booking for an Airbnb, we decided to delete those records.

In [18]:
# Delete the matching documents
result = db.listing_customer.delete_many(
    {
        "$and": [
            {"$or": [{"summary": {"$exists": False}},
                    {"summary": {"$in": [".", " ", "  ", "?", ""]}},]},
            {"$or": [
                    {"description": {"$exists": False}},
                    {"description": {"$in": [".", " ", "  ", "?", ""]}},]},]})

# Print the number of documents that were deleted
print(f"{result.deleted_count} documents deleted.")

95 documents deleted.


##### 2.6. "notes" field

In [19]:
empty_notes = db.listing_customer.aggregate([
    # Find documents where the notes field does not exist or is null
    {"$match": {"$or": [{"notes": {"$exists": False}}, {"notes": {"$in": [".", " ", "  ", "?",""]}}]}},
    # Project the notes field
    {"$project": {"notes": 1}}
])

# Check the count
results = list(empty_notes)
pprint(results)

[{'_id': '1001265', 'notes': ''},
 {'_id': '10021707', 'notes': ''},
 {'_id': '10030955', 'notes': ''},
 {'_id': '10038496', 'notes': ''},
 {'_id': '10047964', 'notes': ''},
 {'_id': '10057447', 'notes': ''},
 {'_id': '10057826', 'notes': ''},
 {'_id': '10059244', 'notes': ''},
 {'_id': '10059872', 'notes': ''},
 {'_id': '10066928', 'notes': ''},
 {'_id': '10069642', 'notes': ''},
 {'_id': '10082307', 'notes': ''},
 {'_id': '10082422', 'notes': ''},
 {'_id': '10083468', 'notes': ''},
 {'_id': '10096773', 'notes': ''},
 {'_id': '10115921', 'notes': ''},
 {'_id': '10116256', 'notes': ''},
 {'_id': '10116578', 'notes': ''},
 {'_id': '10117617', 'notes': ''},
 {'_id': '10120414', 'notes': ''},
 {'_id': '10133350', 'notes': ''},
 {'_id': '10133554', 'notes': ''},
 {'_id': '10140368', 'notes': ''},
 {'_id': '10166883', 'notes': ''},
 {'_id': '10184012', 'notes': ''},
 {'_id': '10186755', 'notes': ''},
 {'_id': '10201975', 'notes': ''},
 {'_id': '10220130', 'notes': ''},
 {'_id': '1022200', '

There are documents where the field `notes` is empty. We considered this to be normal, so we will not delete these documents.

##### 2.7. "access" field

Check the duplicates.

In [20]:
# Check for Duplicates in the field access
duplicates_access = db.listingsAndReviews_new.aggregate(
    # Grouping by access field and counting the number of ocurrences
    [{"$group": {"_id": "$access", "Count": {"$sum": 1}}},
    # Show only if there are duplicates (greater than 1)
    {"$match": {"Count": {"$gt": 1}}}])

# Check the count 
results = list(duplicates_access)
print(results)

[{'_id': "The terrace of Yurbban Passage Hotel is the ideal location for shootings and filmings, a surprise party for your best friend or a laid-back work meeting. All this at the top of the new 'Soho' of Barcelona with one of the best views of the city.", 'Count': 2}, {'_id': 'Cozinha equipada pátio exterior quarto de banho para duches', 'Count': 2}, {'_id': 'You can make use of the full apartment with no restrictions. We will also provide you with all the information you need to know from nearby the apartment so you can feel as a local!', 'Count': 2}, {'_id': 'Please refer the bottom section for english ++++++交通の便利+++++ •ｻﾞﾊﾞｽ(市ﾊﾞｽ) $2.75 (十九､二十番)で乗り換え無し､約1時間でｱﾊﾟｰﾄに到着です｡ ﾊﾞｽ内にお釣りはないのでご自身で用意するのをお忘れなく｡ 大きなｽｰﾂｹｰｽをお持ちの方は乗車出来ませんが､機内持ち込みｻｲｽﾞで来られる方はﾊﾞｽで来れたら格安です! ﾊﾞｽは寒いので上着必須です｡ 空港発ｱﾊﾟｰﾄ前行き(7:00-8:30) ｱﾊﾟｰﾄ前発 空港行き (15:30-18:00) 通勤ﾗｯｼｭなので確実に1時間以上はかかります｡私もこの時間は運転しないように気をつけています｡ 空港発ｱﾊﾟｰﾄ行き降車  "Kuhio Ave and launiu street"  ｸﾋｵｱﾍﾞﾆｭｰ ｱﾝﾄﾞ ﾗｳﾆｳｽﾄﾘｰﾄ (ﾘｯﾂｶｰﾙﾄﾝ目の前) 運転手にﾘｯﾂｶｰﾙﾄﾝに行きたいと伝えたらｽﾑｰｽﾞです｡目の前

##### 2.8. "amenities" field

Check if the amenities field is empty. It's not very common for customers to book an Airbnb without knowing what will be available for them.

In [21]:
empty_amenities = db.listing_customer.aggregate([
    # Find documents where the amenities and description fields are empty or null
    {"$match": {"$or": [{"amenities": {"$exists": False}}, {"amenities": {"$in": [".", " ", "  ", "?",""]}}]}},
    # Project the amenities and description fields
    {"$project": {"Amenities": 1}}
])

# Check the count
results = list(empty_amenities)
pprint(results)

[{'_id': '1077686'},
 {'_id': '11863943'},
 {'_id': '1290112'},
 {'_id': '13145113'},
 {'_id': '13249697'},
 {'_id': '13837488'},
 {'_id': '14554642'},
 {'_id': '16622489'},
 {'_id': '19592057'},
 {'_id': '20354056'},
 {'_id': '20517090'},
 {'_id': '20744948'},
 {'_id': '2436700'},
 {'_id': '2764689'},
 {'_id': '2794928'},
 {'_id': '28002020'},
 {'_id': '2868278'},
 {'_id': '2873436'},
 {'_id': '2946022'},
 {'_id': '2964930'},
 {'_id': '31548152'},
 {'_id': '3786402'},
 {'_id': '4573398'},
 {'_id': '662403'},
 {'_id': '6663848'},
 {'_id': '7080271'},
 {'_id': '7242041'},
 {'_id': '757385'},
 {'_id': '781631'},
 {'_id': '9677123'}]


In [22]:
# Using the delete_many funtion
result = db.listing_customer.delete_many({
    "$or": [{"amenities": {"$exists": False}}, {"amenities": {"$in": [".", " ", "  ", "?",""]}}]
})
# Print the number of documents that were deleted
print(result.deleted_count, "documents deleted.")

30 documents deleted.


In [23]:
# Checking the unique amenities that exist
result = db.listing_customer.aggregate([
    # Unwind the field 'amenities'
    {"$unwind": "$amenities" },
    # Group by the field 'amenities' and count the number of ocurrences
    {"$group": {"_id":"$amenities","Count":{"$sum":1}}},
    # Sort the amenities by the number of ocurrences in descending order
    {"$sort":{ "Count": -1}},
    # Project the 'amenity' and 'Count' fields
    {"$project":{"_id":0, "amenity":"$_id", "Count":1}}
])

pprint(list(result))

[{'Count': 5124, 'amenity': 'Wifi'},
 {'Count': 4875, 'amenity': 'Essentials'},
 {'Count': 4800, 'amenity': 'Kitchen'},
 {'Count': 4146, 'amenity': 'TV'},
 {'Count': 4086, 'amenity': 'Hangers'},
 {'Count': 3769, 'amenity': 'Hair dryer'},
 {'Count': 3760, 'amenity': 'Washer'},
 {'Count': 3578, 'amenity': 'Iron'},
 {'Count': 3573, 'amenity': 'Shampoo'},
 {'Count': 3338, 'amenity': 'Laptop friendly workspace'},
 {'Count': 3318, 'amenity': 'Air conditioning'},
 {'Count': 3184, 'amenity': 'Heating'},
 {'Count': 2862, 'amenity': 'Hot water'},
 {'Count': 2800, 'amenity': 'Smoke detector'},
 {'Count': 2435, 'amenity': 'Family/kid friendly'},
 {'Count': 2327, 'amenity': 'Dryer'},
 {'Count': 2260, 'amenity': 'Elevator'},
 {'Count': 2136, 'amenity': 'Fire extinguisher'},
 {'Count': 1846, 'amenity': 'Refrigerator'},
 {'Count': 1779, 'amenity': 'Internet'},
 {'Count': 1751, 'amenity': 'First aid kit'},
 {'Count': 1718, 'amenity': 'Bed linens'},
 {'Count': 1694, 'amenity': 'Cable TV'},
 {'Count': 16

There are amenities that say "translation missing: en.hosting_amenity_50" and "translation missing: en.hosting_amenity_49", which we will delete.

In [24]:
# Deleting the elements specified above in the amenities array across all documents
db.listing_customer.update_many(
    # Specifying the filter that selects the documents that contain the elements we want to remove
    {"amenities":{"$in":["translation missing: en.hosting_amenity_50", "translation missing: en.hosting_amenity_49"]}},
    # Using the $pull operator to remove the specified elements
    {"$pull":{"amenities":{"$in":["translation missing: en.hosting_amenity_50", "translation missing: en.hosting_amenity_49"]}}}
)

##### 2.9. "accommodates" field 

In [25]:
value_counts = db.listing_customer.aggregate([
    # Grouping by accommodates field and counting the number of ocurrences
    {"$group": {"_id": "$accommodates", "Count": {"$sum": 1}}},
    # Sort the results by ascending order of the "accommodates" field
    {"$sort": {"_id": 1}},
    {"$project": {"Accommodates": "$_id", "Count": 1, "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Accommodates': 1, 'Count': 518},
 {'Accommodates': 2, 'Count': 1971},
 {'Accommodates': 3, 'Count': 543},
 {'Accommodates': 4, 'Count': 1123},
 {'Accommodates': 5, 'Count': 258},
 {'Accommodates': 6, 'Count': 486},
 {'Accommodates': 7, 'Count': 77},
 {'Accommodates': 8, 'Count': 188},
 {'Accommodates': 9, 'Count': 29},
 {'Accommodates': 10, 'Count': 63},
 {'Accommodates': 11, 'Count': 7},
 {'Accommodates': 12, 'Count': 30},
 {'Accommodates': 13, 'Count': 6},
 {'Accommodates': 14, 'Count': 11},
 {'Accommodates': 15, 'Count': 3},
 {'Accommodates': 16, 'Count': 22}]


##### 2.10. "bathrooms" field

In [26]:
value_counts = db.listing_customer.aggregate([
    # Grouping by bathrooms field and counting the number of ocurrences
    {"$group": {"_id": "$bathrooms", "count": {"$sum": 1}}},
    # Sort the results by ascending order of the "bathrooms" field
    {"$sort": {"_id": 1}},
    {"$project": {"Bathrooms": "$_id", "Count": "$count", "_id":0}}
])

# Print the output using pprint
pprint(list(value_counts))

[{'Bathrooms': None, 'Count': 6},
 {'Bathrooms': Decimal128('0.5'), 'Count': 14},
 {'Bathrooms': Decimal128('1.0'), 'Count': 4041},
 {'Bathrooms': Decimal128('1.5'), 'Count': 228},
 {'Bathrooms': Decimal128('2.0'), 'Count': 746},
 {'Bathrooms': Decimal128('2.5'), 'Count': 76},
 {'Bathrooms': Decimal128('3.0'), 'Count': 130},
 {'Bathrooms': Decimal128('3.5'), 'Count': 25},
 {'Bathrooms': Decimal128('4.0'), 'Count': 31},
 {'Bathrooms': Decimal128('4.5'), 'Count': 8},
 {'Bathrooms': Decimal128('5.0'), 'Count': 11},
 {'Bathrooms': Decimal128('5.5'), 'Count': 2},
 {'Bathrooms': Decimal128('6.0'), 'Count': 4},
 {'Bathrooms': Decimal128('7.0'), 'Count': 6},
 {'Bathrooms': Decimal128('8.0'), 'Count': 5},
 {'Bathrooms': Decimal128('9.0'), 'Count': 1},
 {'Bathrooms': Decimal128('16.0'), 'Count': 1}]


There are decimals values for the number of bathrooms. This is possible since, 0.5 could represent an "incomplete bathroom", meaning it doesn't have a shower or bathtub for example.

As for the ``None`` values, we will consider them as 0.

In [27]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$bathrooms", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "bathrooms" field
  {"$sort": {"_id": 1}},
  {"$project": {"Bathrooms": {"$ifNull": ["$_id", 0]}, "Count": "$count", "_id":0}}
])

pprint(list(value_counts))

[{'Bathrooms': 0, 'Count': 6},
 {'Bathrooms': Decimal128('0.5'), 'Count': 14},
 {'Bathrooms': Decimal128('1.0'), 'Count': 4041},
 {'Bathrooms': Decimal128('1.5'), 'Count': 228},
 {'Bathrooms': Decimal128('2.0'), 'Count': 746},
 {'Bathrooms': Decimal128('2.5'), 'Count': 76},
 {'Bathrooms': Decimal128('3.0'), 'Count': 130},
 {'Bathrooms': Decimal128('3.5'), 'Count': 25},
 {'Bathrooms': Decimal128('4.0'), 'Count': 31},
 {'Bathrooms': Decimal128('4.5'), 'Count': 8},
 {'Bathrooms': Decimal128('5.0'), 'Count': 11},
 {'Bathrooms': Decimal128('5.5'), 'Count': 2},
 {'Bathrooms': Decimal128('6.0'), 'Count': 4},
 {'Bathrooms': Decimal128('7.0'), 'Count': 6},
 {'Bathrooms': Decimal128('8.0'), 'Count': 5},
 {'Bathrooms': Decimal128('9.0'), 'Count': 1},
 {'Bathrooms': Decimal128('16.0'), 'Count': 1}]


##### 2.11. "bedrooms" field

In [28]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$bedrooms", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "bedrooms" field
  {"$sort": {"_id": 1}},
  {"$project": {"Bedrooms": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Bedrooms': None, 'Count': 4},
 {'Bedrooms': 0, 'Count': 465},
 {'Bedrooms': 1, 'Count': 3175},
 {'Bedrooms': 2, 'Count': 1057},
 {'Bedrooms': 3, 'Count': 412},
 {'Bedrooms': 4, 'Count': 158},
 {'Bedrooms': 5, 'Count': 34},
 {'Bedrooms': 6, 'Count': 15},
 {'Bedrooms': 7, 'Count': 6},
 {'Bedrooms': 8, 'Count': 3},
 {'Bedrooms': 9, 'Count': 2},
 {'Bedrooms': 10, 'Count': 2},
 {'Bedrooms': 15, 'Count': 1},
 {'Bedrooms': 20, 'Count': 1}]


We'll consider the ``None`` values as 0.

In [29]:
db.listing_customer.update_many({"bedrooms": None }, {"$set": {"bedrooms":0}})

value_counts = db.listing_customer.aggregate([
    # Group by the bedrooms field and count the number of ocurrences
    {"$group": {"_id": "$bedrooms", "Count": {"$sum": 1}}},
    # Sort the results by ascending order of the "bedrooms" field
    {"$sort": {"_id": 1}},
    {"$project": {"Bedrooms": "$_id", "Count": 1, "_id":0}}
])

pprint(list(value_counts))

[{'Bedrooms': 0, 'Count': 469},
 {'Bedrooms': 1, 'Count': 3175},
 {'Bedrooms': 2, 'Count': 1057},
 {'Bedrooms': 3, 'Count': 412},
 {'Bedrooms': 4, 'Count': 158},
 {'Bedrooms': 5, 'Count': 34},
 {'Bedrooms': 6, 'Count': 15},
 {'Bedrooms': 7, 'Count': 6},
 {'Bedrooms': 8, 'Count': 3},
 {'Bedrooms': 9, 'Count': 2},
 {'Bedrooms': 10, 'Count': 2},
 {'Bedrooms': 15, 'Count': 1},
 {'Bedrooms': 20, 'Count': 1}]


##### 2.12. "bed" field

In [30]:
value_counts = db.listing_customer.aggregate([
    # Group by the beds field and count the number of ocurrences
    {"$group": {"_id": "$beds", "count": {"$sum": 1}}},
    # Sort the results by ascending order of the "bed" field
    {"$sort": {"_id": 1}},
    {"$project": {"beds": "$_id", "Count": "$count", "_id":0}}
])

# We print the output using pprint
pprint(list(value_counts))

[{'Count': 9, 'beds': None},
 {'Count': 2601, 'beds': 1},
 {'Count': 1339, 'beds': 2},
 {'Count': 633, 'beds': 3},
 {'Count': 376, 'beds': 4},
 {'Count': 163, 'beds': 5},
 {'Count': 103, 'beds': 6},
 {'Count': 42, 'beds': 7},
 {'Count': 27, 'beds': 8},
 {'Count': 11, 'beds': 9},
 {'Count': 15, 'beds': 10},
 {'Count': 3, 'beds': 11},
 {'Count': 4, 'beds': 12},
 {'Count': 1, 'beds': 13},
 {'Count': 2, 'beds': 14},
 {'Count': 2, 'beds': 15},
 {'Count': 2, 'beds': 16},
 {'Count': 1, 'beds': 18},
 {'Count': 1, 'beds': 25}]


We'll consider the ``None`` values as 0.

In [31]:
db.listing_customer.update_many({"beds": None }, {"$set": {"beds":0}})

value_counts = db.listing_customer.aggregate([
    # Group by the beds field and count the number of ocurrences
    {"$group": {"_id": "$beds", "Count": {"$sum": 1}}},
    # Sort the results by ascending order of the "beds" field
    {"$sort": {"_id": 1}},
    {"$project": {"Beds": "$_id", "Count": 1, "_id":0}}
])

pprint(list(value_counts))

[{'Beds': 0, 'Count': 9},
 {'Beds': 1, 'Count': 2601},
 {'Beds': 2, 'Count': 1339},
 {'Beds': 3, 'Count': 633},
 {'Beds': 4, 'Count': 376},
 {'Beds': 5, 'Count': 163},
 {'Beds': 6, 'Count': 103},
 {'Beds': 7, 'Count': 42},
 {'Beds': 8, 'Count': 27},
 {'Beds': 9, 'Count': 11},
 {'Beds': 10, 'Count': 15},
 {'Beds': 11, 'Count': 3},
 {'Beds': 12, 'Count': 4},
 {'Beds': 13, 'Count': 1},
 {'Beds': 14, 'Count': 2},
 {'Beds': 15, 'Count': 2},
 {'Beds': 16, 'Count': 2},
 {'Beds': 18, 'Count': 1},
 {'Beds': 25, 'Count': 1}]


##### 2.13. "extra_people" field

In [32]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$extra_people", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "bed" field
  {"$sort": {"_id": 1}},
  {"$project": {"extra_people": "$_id", "Count": "$count", "_id":0}}
])

# We print the output using pprint
pprint(list(value_counts))

[{'Count': 2981, 'extra_people': Decimal128('0.00')},
 {'Count': 10, 'extra_people': Decimal128('4.00')},
 {'Count': 57, 'extra_people': Decimal128('5.00')},
 {'Count': 8, 'extra_people': Decimal128('6.00')},
 {'Count': 36, 'extra_people': Decimal128('7.00')},
 {'Count': 24, 'extra_people': Decimal128('8.00')},
 {'Count': 13, 'extra_people': Decimal128('9.00')},
 {'Count': 289, 'extra_people': Decimal128('10.00')},
 {'Count': 5, 'extra_people': Decimal128('11.00')},
 {'Count': 30, 'extra_people': Decimal128('12.00')},
 {'Count': 6, 'extra_people': Decimal128('13.00')},
 {'Count': 4, 'extra_people': Decimal128('14.00')},
 {'Count': 245, 'extra_people': Decimal128('15.00')},
 {'Count': 1, 'extra_people': Decimal128('16.00')},
 {'Count': 10, 'extra_people': Decimal128('17.00')},
 {'Count': 11, 'extra_people': Decimal128('18.00')},
 {'Count': 10, 'extra_people': Decimal128('19.00')},
 {'Count': 272, 'extra_people': Decimal128('20.00')},
 {'Count': 4, 'extra_people': Decimal128('21.00')},
 

##### 2.14. "guests_included" field

In [33]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$guests_included", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "bed" field
  {"$sort": {"_id": 1}},
  {"$project": {"guests_included": "$_id", "Count": "$count", "_id":0}}
])

# We print the output using pprint
pprint(list(value_counts))

[{'Count': 3426, 'guests_included': Decimal128('1')},
 {'Count': 1127, 'guests_included': Decimal128('2')},
 {'Count': 152, 'guests_included': Decimal128('3')},
 {'Count': 365, 'guests_included': Decimal128('4')},
 {'Count': 57, 'guests_included': Decimal128('5')},
 {'Count': 125, 'guests_included': Decimal128('6')},
 {'Count': 14, 'guests_included': Decimal128('7')},
 {'Count': 31, 'guests_included': Decimal128('8')},
 {'Count': 6, 'guests_included': Decimal128('9')},
 {'Count': 18, 'guests_included': Decimal128('10')},
 {'Count': 7, 'guests_included': Decimal128('12')},
 {'Count': 2, 'guests_included': Decimal128('13')},
 {'Count': 2, 'guests_included': Decimal128('15')},
 {'Count': 3, 'guests_included': Decimal128('16')}]


##### 2.15. "host_identity_verified" field

In [34]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$host_identity_verified", "count": {"$sum": 1}}},
  {"$project": {"Host identity verified": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Count': 1946, 'Host identity verified': True},
 {'Count': 3389, 'Host identity verified': False}]


##### 2.16. "host_is_superhost" field

In [35]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$host_is_superhost", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "host.host_is_superhost" field
  {"$sort": {"_id": 1}},
  {"$project": {"Host is Superhost": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Count': 4261, 'Host is Superhost': False},
 {'Count': 1074, 'Host is Superhost': True}]


##### 2.17. "host_listings_count" field

In [36]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$host_listings_count", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "host.host_listings_count" field
  {"$sort": {"_id": 1}},
  {"$project": {"Host Number of Listings": "$_id", "Count": "$count", "_id":0}}
])

# We print the output using pprint
pprint(list(value_counts))

[{'Count': 5, 'Host Number of Listings': 0},
 {'Count': 2419, 'Host Number of Listings': 1},
 {'Count': 790, 'Host Number of Listings': 2},
 {'Count': 368, 'Host Number of Listings': 3},
 {'Count': 233, 'Host Number of Listings': 4},
 {'Count': 160, 'Host Number of Listings': 5},
 {'Count': 138, 'Host Number of Listings': 6},
 {'Count': 108, 'Host Number of Listings': 7},
 {'Count': 88, 'Host Number of Listings': 8},
 {'Count': 70, 'Host Number of Listings': 9},
 {'Count': 49, 'Host Number of Listings': 10},
 {'Count': 47, 'Host Number of Listings': 11},
 {'Count': 59, 'Host Number of Listings': 12},
 {'Count': 41, 'Host Number of Listings': 13},
 {'Count': 29, 'Host Number of Listings': 14},
 {'Count': 32, 'Host Number of Listings': 15},
 {'Count': 28, 'Host Number of Listings': 16},
 {'Count': 30, 'Host Number of Listings': 17},
 {'Count': 26, 'Host Number of Listings': 18},
 {'Count': 22, 'Host Number of Listings': 19},
 {'Count': 18, 'Host Number of Listings': 20},
 {'Count': 15, '

##### 2.18. "host_name" field

In [37]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$host_name", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "host_name" field
  {"$sort": {"_id": 1}},
  {"$project": {"Host Name": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Count': 1, 'Host Name': '(Email hidden by Airbnb)'},
 {'Count': 1, 'Host Name': '(Joe)  American Club'},
 {'Count': 1, 'Host Name': '08028 Apartments'},
 {'Count': 1, 'Host Name': '194 Porto.Flats'},
 {'Count': 1, 'Host Name': '24 Eight Guesthouse'},
 {'Count': 1, 'Host Name': "3M'S - City & Country Rentals"},
 {'Count': 1, 'Host Name': 'A J'},
 {'Count': 1, 'Host Name': 'AFI Apartments'},
 {'Count': 3, 'Host Name': 'Aaron'},
 {'Count': 1, 'Host Name': 'Abagail'},
 {'Count': 1, 'Host Name': 'Abby'},
 {'Count': 2, 'Host Name': 'Abdou'},
 {'Count': 1, 'Host Name': 'Abdul'},
 {'Count': 1, 'Host Name': 'Abdulmutalip'},
 {'Count': 1, 'Host Name': 'Abe'},
 {'Count': 1, 'Host Name': 'Abel'},
 {'Count': 2, 'Host Name': 'Acomodis Apartments'},
 {'Count': 1, 'Host Name': 'Ad'},
 {'Count': 1, 'Host Name': 'Ada'},
 {'Count': 7, 'Host Name': 'Adam'},
 {'Count': 1, 'Host Name': 'Ade'},
 {'Count': 1, 'Host Name': 'Adelle'},
 {'Count': 1, 'Host Name': 'Adelyna'},
 {'Count': 1, 'Host Name': 'Adem'},

##### 2.19. "price" field

In [38]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$price", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "price" field
  {"$sort": {"_id": 1}},
  {"$project": {"Price ": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Count': 1, 'Price ': Decimal128('9.00')},
 {'Count': 2, 'Price ': Decimal128('10.00')},
 {'Count': 2, 'Price ': Decimal128('12.00')},
 {'Count': 3, 'Price ': Decimal128('13.00')},
 {'Count': 5, 'Price ': Decimal128('14.00')},
 {'Count': 17, 'Price ': Decimal128('15.00')},
 {'Count': 7, 'Price ': Decimal128('16.00')},
 {'Count': 10, 'Price ': Decimal128('17.00')},
 {'Count': 7, 'Price ': Decimal128('18.00')},
 {'Count': 5, 'Price ': Decimal128('19.00')},
 {'Count': 24, 'Price ': Decimal128('20.00')},
 {'Count': 7, 'Price ': Decimal128('21.00')},
 {'Count': 9, 'Price ': Decimal128('22.00')},
 {'Count': 11, 'Price ': Decimal128('23.00')},
 {'Count': 8, 'Price ': Decimal128('24.00')},
 {'Count': 47, 'Price ': Decimal128('25.00')},
 {'Count': 8, 'Price ': Decimal128('26.00')},
 {'Count': 9, 'Price ': Decimal128('27.00')},
 {'Count': 10, 'Price ': Decimal128('28.00')},
 {'Count': 11, 'Price ': Decimal128('29.00')},
 {'Count': 62, 'Price ': Decimal128('30.00')},
 {'Count': 13, 'Price ': De

##### 2.20. "property_type" field

In [39]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$property_type", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "property_type" field
  {"$sort": {"_id": 1}},
  {"$project": {"Property Type ": "$_id", "Count": "$count", "_id":0}}
])


# We print the output using pprint
pprint(list(value_counts))

[{'Count': 23, 'Property Type ': 'Aparthotel'},
 {'Count': 3501, 'Property Type ': 'Apartment'},
 {'Count': 1, 'Property Type ': 'Barn'},
 {'Count': 62, 'Property Type ': 'Bed and breakfast'},
 {'Count': 2, 'Property Type ': 'Boat'},
 {'Count': 49, 'Property Type ': 'Boutique hotel'},
 {'Count': 11, 'Property Type ': 'Bungalow'},
 {'Count': 15, 'Property Type ': 'Cabin'},
 {'Count': 1, 'Property Type ': 'Camper/RV'},
 {'Count': 1, 'Property Type ': 'Campsite'},
 {'Count': 8, 'Property Type ': 'Casa particular (Cuba)'},
 {'Count': 1, 'Property Type ': 'Castle'},
 {'Count': 2, 'Property Type ': 'Chalet'},
 {'Count': 382, 'Property Type ': 'Condominium'},
 {'Count': 20, 'Property Type ': 'Cottage'},
 {'Count': 1, 'Property Type ': 'Earth house'},
 {'Count': 9, 'Property Type ': 'Farm stay'},
 {'Count': 77, 'Property Type ': 'Guest suite'},
 {'Count': 46, 'Property Type ': 'Guesthouse'},
 {'Count': 1, 'Property Type ': 'Heritage hotel (India)'},
 {'Count': 29, 'Property Type ': 'Hostel'},


##### 2.21. "room_type" field

In [40]:
value_counts = db.listing_customer.aggregate([
  {"$group": {"_id": "$room_type", "count": {"$sum": 1}}},
  # Sort the results by ascending order of the "room_type" field
  {"$sort": {"_id": 1}},
  {"$project": {"Room Type ": "$_id", "Count": "$count", "_id":0}}
])

# We print the output using pprint
pprint(list(value_counts))

[{'Count': 3386, 'Room Type ': 'Entire home/apt'},
 {'Count': 1873, 'Room Type ': 'Private room'},
 {'Count': 76, 'Room Type ': 'Shared room'}]


#### 3. The number of guests included is higher than the number of accomodates an Airbnb accomodates

In [41]:
guests_greater_accomodates = db.listing_customer.aggregate([
    # Using the greater than operator
  { "$match": { "guests_included": { "$lt": "accommodates" } } },
  { "$count": "total" }
])

# Check the count 
results = list(guests_greater_accomodates)
print(results)



[]


In [42]:
missing = db.listing_customer.aggregate([
  { "$match": { "price": { "$exists":False , "$eq": False } } }
])

# Check the count 
results = list(missing)
print(results)

[]


### Question 3 
Once per month we like to reward hosts with recognition. Pick three superhosts who have at least two property listings that can accommodate more than four people?

In [43]:
# Store the 3 hosts in a variable called three_superhosts

three_superhosts = list( db.listing_customer.aggregate([
        # Specify the conditions
        {"$match": {"host_is_superhost": True, # Being a superhost
                    "host_listings_count": {"$gte" :2 }, # Having at least 2 property listings
                    "accommodates": {"$gt" : 4}}}, # Accommodate more than 4 people
        # Limiting to 3 superhosts
         {"$limit":3},
    
        # Choose what we want to return
         {"$project": {"ID of the Host": "$_id", # Retun the _id as "ID of the Host"
                       "Name of the Host":"$host_name", # Return the host_name as "Name of the Host"
                       "Number of Properties":"$host_listings_count"}}])) # Return the nr. of properties as "Number of Properties"

# Pretty print the 3 superhosts to reward
pprint(three_superhosts)

[{'ID of the Host': '10548991',
  'Name of the Host': 'Nate',
  'Number of Properties': 3,
  '_id': '10548991'},
 {'ID of the Host': '10570041',
  'Name of the Host': 'Roberta (Beta) Gatti',
  'Number of Properties': 18,
  '_id': '10570041'},
 {'ID of the Host': '10635276',
  'Name of the Host': 'Luiz Henrique',
  'Number of Properties': 2,
  '_id': '10635276'}]


### Question 4 
One of employees is thinking of buying a property to rent out. Which bed type is most common in the listings that have waterfront and a dishwasher in New York?

In [44]:
bed_type = list(db.listing_customer.aggregate([
    # We use a $lookup to join the 'listing_customer' and 'listing_reviews' collections
    {"$lookup": {
        "from": "listing_reviews",
        "localField": "_id", # Field in the listing_address we want to match
        "foreignField": "listing_id", # Field in the listing_reviews we want to match
        "as": "review_info" #Giving an alias 
    }},
    # Specifying the conditions
    {"$match": {"address.market": "New York",  # In New York
                "amenities": {"$all": ["Waterfront", "Dishwasher"]}}}, # Having a dishwasher and a waterfront
    # Grouping by bed_type
    {"$group": {"_id": "$bed_type",
                "count": {"$sum": 1}}},
    # Sort in a descending order
    {"$sort": {"count": -1}},
    # Limiting to the most common bed type
    {"$limit": 1},
    # Returning the bed type name
    {"$project": {"_id": 0,
                  "Bed type name": "$_id"}}
]))

# Print the result
pprint(bed_type)

[{'Bed type name': 'Real Bed'}]


### Question 5
We are thinking of identifying someone to hire to write review for us professionally. What is the name of the reviewer who left the longest review in New York?

In [45]:
# https://www.mongodb.com/docs/manual/reference/operator/aggregation/strLenCP/
# What is the name of the reviewer who left the longest review in New York?
reviewer_ny = list(db.listing_reviews.aggregate([
    # Using lookup to join the "listing_reviews" and "listing_customer" collections
    {"$lookup": {
        "from": "listing_customer",
        "localField": "listing_id",
        "foreignField": "_id",
        "as": "customer"
    }},
    # We use the match to filter the documents in the collection from people that are from New York
    {"$match": {"customer.address.market": "New York",
                "comments": {"$exists": True}}},
    # $addFields adds a new field named 'review_length' to each review document
    # We use the operator $strLenCP to calculate the length of the comments
    {"$addFields": {"review_length": {"$strLenCP": "$comments"}}},
    # We then sort the documents in descending order of review_length, so that the longest review appears first
    {"$sort": {"review_length": -1}},
    {"$limit": 1},
    {"$project": {"_id": 0, "Reviewer Name": "$reviewer_name"}}
]))

pprint(reviewer_ny)

[{'Reviewer Name': 'Angela'}]


On question 6 and 7 we use the initial collection as the new collections were making the queries very slow. Still we left the query of 6 below in commentary.

### Question 6
We are thinking of trying to assess the security of different areas based on the security deposit required. What is the biggest and smallest difference between the price and deposit for security per number of visitors staying at the property?

In [46]:
dif = list(db.listingsAndReviews_new.aggregate([
    # We want the documents where we have at least 1 guest and where security_deposit exists
    {"$match": {"$and": [{"security_deposit": {"$exists": True}}, {"accommodates": {"$gte": 1}}]}},
    # We project the fields we need to calculate the difference and the difference itself 
    {"$project": {
        "_id": 0, 
        "accommodates": 1, 
        "price": 1, 
        "security_deposit": 1, 
        "difference": {"$abs": {"$subtract": ["$price", "$security_deposit"]}},
        "address.city": 1, 
        "reviews": 1
    }},
    # Then we group by the number of guests and calculate the max and min difference
    {"$group": {"_id": "$accommodates", "max_difference": {"$max": "$difference"}, "min_difference": {"$min": "$difference"}}},
    # Change the name of _id to accommodates
    {"$project": {"_id": 0, "Number of accommodates": "$_id", "Maximum difference":"$max_difference", "Minimum difference":"$min_difference"}},
    # Sort by the number of guests
    {"$sort": {"accommodates": 1}}
]))

pprint(dif)

[{'Maximum difference': Decimal128('8501.00'),
  'Minimum difference': Decimal128('1.00'),
  'Number of accommodates': 5},
 {'Maximum difference': Decimal128('2501.00'),
  'Minimum difference': Decimal128('0.00'),
  'Number of accommodates': 7},
 {'Maximum difference': Decimal128('19049.00'),
  'Minimum difference': Decimal128('1.00'),
  'Number of accommodates': 8},
 {'Maximum difference': Decimal128('1350.00'),
  'Minimum difference': Decimal128('8.00'),
  'Number of accommodates': 9},
 {'Maximum difference': Decimal128('14631.00'),
  'Minimum difference': Decimal128('0.00'),
  'Number of accommodates': 1},
 {'Maximum difference': Decimal128('29148.00'),
  'Minimum difference': Decimal128('0.00'),
  'Number of accommodates': 4},
 {'Maximum difference': Decimal128('10498.00'),
  'Minimum difference': Decimal128('80.00'),
  'Number of accommodates': 16},
 {'Maximum difference': Decimal128('38459.00'),
  'Minimum difference': Decimal128('0.00'),
  'Number of accommodates': 2},
 {'Maximu

In [47]:
#dif = list(db.listing_customer.aggregate([
#    # We want the documents where we have at least 1 guest and where security_deposit exists
#    {"$match": {"$and": [{"security_deposit": {"$exists": True}}, {"accommodates": {"$gte": 1}}]}},
#    # We join with "listing_reviews" collections to get the necessary fields
#    {"$lookup": {
#        "from": "listing_reviews",
#        "localField": "_id",
#        "foreignField": "listing_id",
#        "as": "reviews"
#    }},
#    # We project the fields we need to calculate the difference and the difference itself 
#    {"$project": {
#        "_id": 0, 
#        "accommodates": 1, 
#        "price": 1, 
#        "security_deposit": 1, 
#        "difference": {"$abs": {"$subtract": ["$price", "$security_deposit"]}},
#        "address.city": 1, 
#        "reviews": 1
#    }},
#    # Then we group by the number of guests and calculate the max and min difference
#    {"$group": {"_id": "$accommodates", "max_difference": {"$max": "$difference"}, "min_difference": {"$min": "$difference"}}},
#    # Change the name of _id to accommodates
#    {"$project": {"_id": 0, "Number of accommodates": "$_id", "Maximum difference":"$max_difference", "Minimum difference":"$min_difference"}},
#    # Sort by the number of guests
#    {"$sort": {"accommodates": 1}}
#]))
#
#pprint(dif)

### Question 7 
We want to identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. We will use this information to target advertising of different customers. It is not expected this information will change much over time so we won’t look to update it we just would like a current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

In [48]:
avg_duration = list(db.listingsAndReviews_new.aggregate([
    # The field maximum_nights is a string so we can't use {"maximum_nights": {"$ne": None}} 
    # First we need to convert it into an integer type
    {"$addFields": {"maximum_nights_int": {"$toInt": "$maximum_nights"}}},
    # Now we filter the documents where maximum_nights_int is not none
    {"$match": {"maximum_nights_int": {"$ne": None}}},
    # Group by property type and city and then we calculate the average duration of stay
    {"$group": {"_id": {"property_type": "$property_type", "city": "$address.market"}, "avg_duration": {"$avg": "$maximum_nights_int"}}},
    # Sort by property type and average duration of stay in descending order
    {"$sort": {"_id.property_type": 1, "avg_duration": -1}},
    # Group by property type 
    # The highest average duration of stay is the first in the sorted list
    # The lowest average duration of stay is the last in the sorted list
    {"$group": {"_id": "$_id.property_type", 
                "Highest duration city": {"$first": "$_id.city"},
                "Highest duration": {"$first": "$avg_duration"},
                "Lowest duration city": {"$last": "$_id.city"},
                "Lowest duration": {"$last": "$avg_duration"}}}
]))

pprint(avg_duration)

[{'Highest duration': 1125.0,
  'Highest duration city': 'Barcelona',
  'Lowest duration': 846.0,
  'Lowest duration city': 'The Big Island',
  '_id': 'Farm stay'},
 {'Highest duration': 1125.0,
  'Highest duration city': 'Rio De Janeiro',
  'Lowest duration': 1125.0,
  'Lowest duration city': 'Rio De Janeiro',
  '_id': 'Earth house'},
 {'Highest duration': 1125.0,
  'Highest duration city': 'The Big Island',
  'Lowest duration': 30.0,
  'Lowest duration city': 'Kauai',
  '_id': 'Bed and breakfast'},
 {'Highest duration': 1125.0,
  'Highest duration city': 'Barcelona',
  'Lowest duration': 235.4,
  'Lowest duration city': 'Istanbul',
  '_id': 'Casa particular (Cuba)'},
 {'Highest duration': 1125.0,
  'Highest duration city': 'Maui',
  'Lowest duration': 35.0,
  'Lowest duration city': 'Other (International)',
  '_id': 'Hotel'},
 {'Highest duration': 1125.0,
  'Highest duration city': 'Barcelona',
  'Lowest duration': 1125.0,
  'Lowest duration city': 'Porto',
  '_id': 'Boat'},
 {'Highe

### Question 8
We want to have a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

Since the data will need to be available every time a host registers a property, we decided to create a new collection `listing_amenities` for optimization.

This collection lists all unique amenities currently listed in properties (across all documents) ordered by most common to least common.

In [49]:
# 1. Creating a new collection 'listing_amenities'

db.listing_customer.aggregate([
    # Unwind the field 'amenities'
    {"$unwind": "$amenities"},
    # Group by the field 'amenities' and count the number of ocurrences
    {"$group": {"_id":"$amenities", "Count":{"$sum":1}} },
    # Sort the amenities by the number of ocurrences in descending order
    {"$sort": {"Count": -1} },
    # Project the 'amenity' field
    {"$project": {"_id": 0, "amenity":"$_id"} },
    # Using the $out to create a new collection named "listing_amenities"
    {"$out": "listing_amenities" }
])

To list the suggested typical amenities in the webpage, we considered the first 20 amenities of the collection created above `listing_amenities`.

We decided to use the top 20 amenities as the most common ones instead of using the number of times each amenity occurs across all documents because, as more properties are added to the database, the number of ocurrences will change.

In [50]:
# 2. Listing the typical amenities
typical_amenities = list(db.listing_amenities.aggregate([
    # Project the 'amenity' field
    {"$project":{"_id": 0, "amenity": 1}},
    # Limit to the first 20 amenities
    {"$limit": 10}
]))

# Extracting the values of the 'amenity' field and storing them in a list
amenities_list = [doc['amenity'] for doc in typical_amenities]

# Printing the list
print('Typical Amenities:\n')
for i in amenities_list:
    print(i)

Typical Amenities:

Wifi
Essentials
Kitchen
TV
Hangers
Hair dryer
Washer
Iron
Shampoo
Laptop friendly workspace


### Question 9
We want to start tracking our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

In [51]:
# 1. Showing the top 20 reviewers and the count of the number of reviews of each of these reviewers

top20_reviewers = list(db.listing_reviews.aggregate([
    # Group by reviewer (id and name) and count the number of reviews for each reviewer
    {"$group": {"_id": {"id": "$reviewer_id", "name": "$reviewer_name"}, "Count": {"$sum": 1}}},
    # Sort by number of reviews in descending order
    {"$sort": {"Count": -1}},
    # Using the $project operator to specify the fields 'reviewer_id' and 'reviewer_name' and the number of reviews (Count)
    {"$project": {"_id": "$_id.id", "Reviewer Name": "$_id.name", "Count": 1}},
    # Limit the result to the top 20 reviewers
    {"$limit": 20}
]))

print("Top 20 reviewers:")
for i in top20_reviewers:
            print('Name: ' + str(i['Reviewer Name']) + ', ID: ' + str(i['_id'])+', Reviews: ' + str(i['Count']))

Top 20 reviewers:
Name: Filipe, ID: 20775242, Reviews: 24
Name: Nick, ID: 67084875, Reviews: 13
Name: Uge, ID: 2961855, Reviews: 10
Name: Lisa, ID: 20991911, Reviews: 9
Name: Thien, ID: 162027327, Reviews: 9
Name: David, ID: 1705870, Reviews: 8
Name: Lisa, ID: 69140895, Reviews: 8
Name: Courtney, ID: 55241576, Reviews: 8
Name: Todd, ID: 60816198, Reviews: 8
Name: Jodi, ID: 12679057, Reviews: 8
Name: David, ID: 78093968, Reviews: 7
Name: Lance, ID: 47303133, Reviews: 7
Name: Assis, ID: 76782210, Reviews: 6
Name: Pierre, ID: 93859831, Reviews: 6
Name: Chris, ID: 86665925, Reviews: 6
Name: Branden, ID: 128210181, Reviews: 6
Name: Megan, ID: 25715809, Reviews: 6
Name: Erik, ID: 61469899, Reviews: 6
Name: Karen, ID: 24667379, Reviews: 6
Name: Gonzalo, ID: 73708321, Reviews: 6


In [52]:
# 2. Creating a function that shows the number of reviews for a given reviewer ID or Name

def review_count_by_identity(reviewer_identity):
    """ This function returns the number of reviews for a given reviewer ID or Name. """
    # if the input is an ID
    if str(reviewer_identity).isdigit():
        # Count the number of documents in the collection 'listing_reviews' where the field 'reviewer_id' matches the given ID
        # Using str() in case the user gives an integer instead of a string
        review_count = db.listing_reviews.count_documents({"reviewer_id": str(reviewer_identity)})
        # Print the result
        print('The reviewer with ID '+str(reviewer_identity)+' wrote '+str(review_count)+' reviews.')
        
    # if the input is the name of a reviewer
    else:
        # Create a list with each reviewer with the given name and the respective number of reviews
        reviewers= list(db.listing_reviews.aggregate([
            # Get the documents where the field 'reviewer_name' matches the given name
            {"$match": {"reviewer_name": reviewer_identity}},
            # Group by the field 'reviewer_id' and count the number of reviews
            {"$group": {"_id": "$reviewer_id", "Count": {"$sum": 1}}},
            # Sort by the number of reviews in descending order
            {"$sort": {"Count": -1}},
            # Specify the fields '_id' and 'Count'
            {"$project": {"_id": 1, "Count": 1}}]))
        for i in reviewers:
            print('The reviewer with name '+str(reviewer_identity)+' and ID '+str(i['_id'])+' wrote '+str(i['Count'])+' reviews.')

In [53]:
# Testing the function using a reviewer ID
reviewer_id = 20775242
review_count_by_identity(reviewer_id)

The reviewer with ID 20775242 wrote 24 reviews.


In [54]:
# Testing the function using a reviewer Name
reviewer_name = 'Filipe'
review_count_by_identity(reviewer_name)

The reviewer with name Filipe and ID 20775242 wrote 24 reviews.
The reviewer with name Filipe and ID 216052293 wrote 1 reviews.
The reviewer with name Filipe and ID 59961074 wrote 1 reviews.
The reviewer with name Filipe and ID 33513898 wrote 1 reviews.
The reviewer with name Filipe and ID 4772886 wrote 1 reviews.
The reviewer with name Filipe and ID 36534285 wrote 1 reviews.
The reviewer with name Filipe and ID 6133482 wrote 1 reviews.
The reviewer with name Filipe and ID 159720185 wrote 1 reviews.
The reviewer with name Filipe and ID 58445636 wrote 1 reviews.
The reviewer with name Filipe and ID 44381882 wrote 1 reviews.
The reviewer with name Filipe and ID 128333821 wrote 1 reviews.
The reviewer with name Filipe and ID 93921200 wrote 1 reviews.
The reviewer with name Filipe and ID 8922130 wrote 1 reviews.
The reviewer with name Filipe and ID 200879734 wrote 1 reviews.
The reviewer with name Filipe and ID 30753558 wrote 1 reviews.
The reviewer with name Filipe and ID 1576616 wrote 1 

### Question 10
For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We are thinking we may soon add more metrics, although we don’t know what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

In order to easily query the average score across all of these metrics, we decided to merge the review scores metrics into one embedded field called `review_scores`.

Since the review score "review_scores_rating" is not between 0 and 10, we decided not to include it in the calculation of the average.

This way, by adding a any new metrics to this new field, the average is easily queried.

In [55]:
# Merge the review scores metrics into one embedded field

update = [
    # Create a new 'review_scores' field containing the embedded fields review scores fields
    {"$set": {"review_scores": {"$mergeObjects": [{"checkin": "$review_scores_checkin"},
                                                  {"cleanliness": "$review_scores_cleanliness"},
                                                  {"communication": "$review_scores_communication"},
                                                  {"location": "$review_scores_location"},
                                                  {"value": "$review_scores_value"}]}}},
    # Remove the original review scores fields
    {"$unset": ["review_scores_checkin", "review_scores_cleanliness", "review_scores_communication",
                "review_scores_location", "review_scores_rating", "review_scores_value"]}]

# Update all documents in the collection
db.listingsAndReviews_new.update_many({}, update)

In [56]:
# Calculating the average for an example property

result = list(db.listingsAndReviews_new.aggregate([
    # Filter the property we wish to query
    {"$match": {"_id":'10006546'}},
    # Create a new field 'review_scores_array' by using the
    # $objectToArray operator to convert the 'review_scores' subdocument into an array of key-value pairs
    {"$project":{"review_scores_array": {"$objectToArray": "$review_scores"}}},
    # Unwind the field 'review_scores_array' so that each element of the array represents a field
    {"$unwind": "$review_scores_array"},
    # Group by to calculate the average of the fields
    {"$group": {"_id": "$_id", "avg_review_score": {"$avg": "$review_scores_array.v"}}},
    # Remove the '_id' from the output and include the 'avg_review_score' field
    {"$project": {"_id": 0, "avg_review_score": {"$round": ["$avg_review_score", 2]}}}
]))

pprint(result)

[{'avg_review_score': 9.6}]


### Question 11
We wish to be able to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

In [57]:
# Define the function that calculates the average value of transactions for the specific period of time and property given
def avg_transaction_value(property_id, start_date, end_date):
    """ This function calculates the average value of transactions of a given property and for a given period of time."""
    
    # Define the aggregation pipeline to retrieve the average transaction value
    pipeline = [
        # Unwind the transactions array
        {"$unwind": "$transactions.transactions"},
        # Project the necessary fields and convert the date to a string
        {"$project": {
            "_id": 1,
            "transactions.transactions.date": {"$dateToString": {"format": "%Y, %m, %d, %H, %M", "date": "$transactions.transactions.date"}},
            "transactions.transactions.price": {"$toDouble": "$transactions.transactions.price"}
        }},
        # Filter by the specified property and the specified time period
        {"$match": {"$and": [{"_id": property_id},
                             {"transactions.transactions.date": {"$gte": start_date.strftime("%Y, %m, %d, %H, %M")}},
                             {"transactions.transactions.date": {"$lte": end_date.strftime("%Y, %m, %d, %H, %M")}}]}},
        # Use "_id": None to calculate the average price across all documents
        #(the id is always the same because it was defined at the beginning)
        {"$group": {"_id": None, "avg_price": {"$avg": "$transactions.transactions.price"}}}
    ]
    
    # Execute the aggregation pipeline and store the results in the 'result' variable
    result = list(db.listingsAndReviews_new.aggregate(pipeline))
    
    # If the result set is empty, print an error message
    if len(result) == 0:
        print("There are no documents in the collection that match the query.")
    # If the result set is not empty, print the average transaction price for the given property and time period
    else:
        print(f"The average transaction price for property {property_id} between {start_date.strftime('%Y-%m-%d')} and {end_date.strftime('%Y-%m-%d')} is {result[0]['avg_price']:.2f}.")
    

In [58]:
# Testing the function

from datetime import datetime

# Specify the property of interest
property_id = '10006546'
# Specify the period of time of interest
start_date = datetime(2000, 1, 1, 0, 0)
end_date = datetime(2010, 12, 31, 0, 0)

avg_transaction_value(property_id, start_date, end_date)

The average transaction price for property 10006546 between 2000-01-01 and 2010-12-31 is 37.43.


### Question 12
We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

In [59]:
# Just checking the first 10 cities
top10_cities_ = list(db.listingsAndReviews_new.aggregate([
    # Grouping the documents by the "address.market" field, and creating a count of how many documents there are for each group
  { '$group': { '_id': "$address.market", 'count': { '$sum': 1 } } },
    # Sorting the results by the count value, in a descending order
  { '$sort': { 'count': -1 } },
    # Project only the id and the count
    {"$project": {"_id": 1, "Count": 1}},
    # Limiting the output to 10 documents
  { '$limit': 10 }
])
                   )
pprint(top10_cities_)

[{'_id': 'Istanbul'},
 {'_id': 'Montreal'},
 {'_id': 'Barcelona'},
 {'_id': 'Hong Kong'},
 {'_id': 'Sydney'},
 {'_id': 'New York'},
 {'_id': 'Rio De Janeiro'},
 {'_id': 'Porto'},
 {'_id': 'Oahu'},
 {'_id': 'Maui'}]


In [60]:
# Top 10 cities and their top 3 properties (based on their average rating)
top10_cities = list(db.listingsAndReviews_new.aggregate([
    # First we want to group by cities so that we can find the top 10 using the sort
  { '$group': { '_id': "$address.market", 
               'count': { '$sum': 1 },
              } },
  { '$sort': { 'count': -1 } },
  { '$limit': 10 },
    
    # We use lookup to make a list of all the properties in each city so that we can order them using review ratings
  { '$lookup': { 'from': "listingsAndReviews_new", 
                'localField': "_id", 
                'foreignField': "address.market", 
                'as': "properties" } },
    # We use unwind to have a different document for each property
  { '$unwind': "$properties" },
    # Again so that we can have all the different reviews separated 
  { '$unwind': "$properties.reviews" },
  { '$group': { '_id': "$properties._id", 
               'avg_rating': { '$max': "$properties.review_scores_rating" } } },
    # Sorting the properties by highest review scores rating
  { '$sort': { 'avg_rating': -1 } },
  { '$group': { '_id': "$_id", 
               'avg_rating': { '$first': "$avg_rating" } } },
    # Then in each property we also want a list of informations on that specific property
  { '$lookup': { 'from': "listingsAndReviews_new", 
                'localField': "_id", 
                'foreignField': "_id", 
                'as': "property_info" } },
    # We chose some variables that we thought were more useful for the summary webpage
  { '$project': { 'city': "$property_info.address.market", 
                 'property_id': "$_id",
                 'avg_rating': 1, #The rating of the property
                 'number_of_reviews': '$property_info.number_of_reviews', #Number of reviews 
                 'price':'$property_info.price', #Price
                 # If the host is a superhost, meaning he will probably be trustworthy
                 'host_is_superhost':'$property_info.host_is_superhost',
                 'property_type':'$property_info.property_type', # The type of property
                 'property_name': "$property_info.name" } }, # Property name
    # Grouping the documents by city
  { '$group': { '_id': "$city", 
               'properties': { '$push':  # Creating a new field called properties
                              { 
                                   'Property ID': "$property_id", 
                                   'Name': "$property_name", 
                                   'Rating': "$avg_rating",
                                   'Number of reviews':'$number_of_reviews',
                                   'Superhost':'$host_is_superhost',
                                   'Type of property':'$property_type',
                                   'Price': '$price'} } } },
  { '$project': { '_id': 0, 
                 'City': "$_id", # Making the id the City
                 'Properties': { '$slice': [ "$properties", 3 ] } } } # Top 3 properties
]))

pprint(top10_cities)

[{'City': ['Hong Kong'],
  'Properties': [{'Name': ['Cozy Room at TKL MTR'],
                  'Number of reviews': [6],
                  'Price': [Decimal128('447.00')],
                  'Property ID': '13962683',
                  'Rating': None,
                  'Superhost': [False],
                  'Type of property': ['Apartment']},
                 {'Name': ['D5.Cozy Private Room for 4 Guest(Private '
                           'Bathroom)'],
                  'Number of reviews': [13],
                  'Price': [Decimal128('2002.00')],
                  'Property ID': '18841656',
                  'Rating': None,
                  'Superhost': [False],
                  'Type of property': ['Hotel']},
                 {'Name': ['NEW FLAT + PRIVATE ROOFTOP IN SAY YING PUN'],
                  'Number of reviews': [3],
                  'Price': [Decimal128('903.00')],
                  'Property ID': '11305966',
                  'Rating': None,
                  'Superhost'

In [61]:
#number of each property type per city

result = list(db.listingsAndReviews_new.aggregate([
    # Group the number of properties for each property type by city
    {"$group": {
        "_id": {"market": "$address.market", "property_type": "$property_type"},
        # Count the number of documents in each group
        "count": {"$sum": 1}
    }},
    # We group again by market, but this time we use the $push operator to create an array with the property 
    # type and count for each group. This will create a nested array that lists the number of properties for 
    # each property type in each city
    # https://www.tutorialspoint.com/mongodb-push-in-nested-array
    {"$group": {
        "_id": "$_id.market",
        "Number of properties per type": {
            "$push": {
                "property_type": "$_id.property_type",
                "count": "$count"
            }
        }
    }}
]))
pprint(result)

[{'Number of properties per type': [{'count': 1,
                                     'property_type': 'Condominium'}],
  '_id': 'Other (Domestic)'},
 {'Number of properties per type': [{'count': 466,
                                     'property_type': 'Apartment'},
                                    {'count': 1, 'property_type': 'Camper/RV'},
                                    {'count': 17, 'property_type': 'Townhouse'},
                                    {'count': 2, 'property_type': 'Hostel'},
                                    {'count': 2,
                                     'property_type': 'Guest suite'},
                                    {'count': 1, 'property_type': 'Bungalow'},
                                    {'count': 1, 'property_type': 'Aparthotel'},
                                    {'count': 18,
                                     'property_type': 'Serviced apartment'},
                                    {'count': 1, 'property_type': 'Cabin'},
           

In [62]:
# average price in the city
result = list(db.listingsAndReviews_new.aggregate([
    # Group the listings by city and get the count and average price for each city
    { '$group': { '_id': "$address.market", 'count': { '$sum': 1 }, 'price': { '$avg': "$price" } } },
    # Limit the output to the top 10 cities by average price
    { '$limit': 10 },
    # Project the city name and average price fields
    { '$project': { '_id': 0, 'City': "$_id", 'Average price': '$price' } }
]))

pprint(result)


[{'Average price': Decimal128('367.9454545454545454545454545454545'),
  'City': 'Istanbul'},
 {'Average price': Decimal128('445.75'), 'City': 'Other (International)'},
 {'Average price': Decimal128('128.00'), 'City': 'Other (Domestic)'},
 {'Average price': Decimal128('179.4604316546762589928057553956835'),
  'City': 'The Big Island'},
 {'Average price': Decimal128('100.2330246913580246913580246913580'),
  'City': 'Montreal'},
 {'Average price': Decimal128('197.7142857142857142857142857142857'),
  'City': 'Sydney'},
 {'Average price': Decimal128('69.12815884476534296028880866425993'),
  'City': 'Porto'},
 {'Average price': Decimal128('762.4781906300484652665589660743134'),
  'City': 'Hong Kong'},
 {'Average price': Decimal128('212.2964426877470355731225296442688'),
  'City': 'Oahu'},
 {'Average price': Decimal128('286.5882352941176470588235294117647'),
  'City': 'Maui'}]


In [63]:
# average availability
result = list(db.listingsAndReviews_new.aggregate([
    # Group the listings by city (address.market), and calculate the count of documents and average availability_30
    { '$group': { '_id': "$address.market", 'count': { '$sum': 1 }, 'availability_30': { '$avg': "$availability.availability_30" } } },
    # Limit the results to 10 documents
    { '$limit': 10 },
     #We assume that availability_30 is availability in the next 30 days
    { '$project': { '_id': 0, 'City': "$_id", 'Availability in the next 30 days':'$availability_30' } } 
]))

pprint(result)

[{'Availability in the next 30 days': 8.313014827018122, 'City': 'New York'},
 {'Availability in the next 30 days': 27.75, 'City': 'Other (International)'},
 {'Availability in the next 30 days': 24.0, 'City': 'Other (Domestic)'},
 {'Availability in the next 30 days': 9.243827160493828, 'City': 'Montreal'},
 {'Availability in the next 30 days': 11.18705035971223,
  'City': 'The Big Island'},
 {'Availability in the next 30 days': 7.1330049261083746, 'City': 'Sydney'},
 {'Availability in the next 30 days': 16.73826714801444, 'City': 'Porto'},
 {'Availability in the next 30 days': 11.082390953150242, 'City': 'Hong Kong'},
 {'Availability in the next 30 days': 4.915032679738562, 'City': 'Maui'},
 {'Availability in the next 30 days': 8.861660079051383, 'City': 'Oahu'}]


### Question 13
Add a new property, this property should have a new host, and should be located in one of the top 10 cities. The host selects the top 10 most common amenities to be listed for the property.

In [64]:
import random

# Define a new 'host_id' by adding 1 to the previous maximum 'host_id'
new_host_id = str(int(list(db.listingsAndReviews_new.aggregate([{"$sort": {"host_id": -1}},
                                                        {"$limit":1},
                                                        {"$project":{"_id":0, "host_id":1}}]))[0]['host_id'])+1)

# Extract the cities field and storing them in a list
top10_cities = [doc['_id'] for doc in top10_cities_]
# Randomly select a city from the top 10 cities
city = random.choice(top10_cities)

# Add a new property to the collection 'listingsAndReviews_new'
db.listingsAndReviews_new.insert_one(
    # Creating the field 'amenities' with the top 10 most common amenities listed in Question 8
    {"amenities": amenities_list,
    "host_id": new_host_id,
    "address":{"market": city}}
)

In [65]:
# Check the new document
result=list(db.listingsAndReviews_new.aggregate([{"$match": {"host_id": new_host_id}}]))

pprint(result) 

[{'_id': ObjectId('64468b324ec1031385ce04e4'),
  'address': {'market': 'Rio De Janeiro'},
  'amenities': ['Wifi',
                'Essentials',
                'Kitchen',
                'TV',
                'Hangers',
                'Hair dryer',
                'Washer',
                'Iron',
                'Shampoo',
                'Laptop friendly workspace'],
  'host_id': '99997585'}]


In [66]:
#db.listingsAndReviews_new.delete_many({ "host_id": "99997586" })

### Question 14
Add a new review to this property, the review should be from one of our top 20 reviewers

In [67]:
# Check the top 20 reviewer and chose one
top20_reviewers

[{'Count': 24, '_id': '20775242', 'Reviewer Name': 'Filipe'},
 {'Count': 13, '_id': '67084875', 'Reviewer Name': 'Nick'},
 {'Count': 10, '_id': '2961855', 'Reviewer Name': 'Uge'},
 {'Count': 9, '_id': '20991911', 'Reviewer Name': 'Lisa'},
 {'Count': 9, '_id': '162027327', 'Reviewer Name': 'Thien'},
 {'Count': 8, '_id': '1705870', 'Reviewer Name': 'David'},
 {'Count': 8, '_id': '69140895', 'Reviewer Name': 'Lisa'},
 {'Count': 8, '_id': '55241576', 'Reviewer Name': 'Courtney'},
 {'Count': 8, '_id': '60816198', 'Reviewer Name': 'Todd'},
 {'Count': 8, '_id': '12679057', 'Reviewer Name': 'Jodi'},
 {'Count': 7, '_id': '78093968', 'Reviewer Name': 'David'},
 {'Count': 7, '_id': '47303133', 'Reviewer Name': 'Lance'},
 {'Count': 6, '_id': '76782210', 'Reviewer Name': 'Assis'},
 {'Count': 6, '_id': '93859831', 'Reviewer Name': 'Pierre'},
 {'Count': 6, '_id': '86665925', 'Reviewer Name': 'Chris'},
 {'Count': 6, '_id': '128210181', 'Reviewer Name': 'Branden'},
 {'Count': 6, '_id': '25715809', 'Rev

From the top 20 reviewers we have chosen the reviewer Nick with id 67084875.

In [68]:
property_id = list(db.listingsAndReviews_new.aggregate([
    {"$match": {"host_id": new_host_id}},
    {"$project": {"_id": 1}}
]))[0]["_id"]
property_id

ObjectId('64468b324ec1031385ce04e4')

In [69]:
# find the property by '_id', to add the review to
db.listingsAndReviews_new.find_one({"_id": property_id})

{'_id': ObjectId('64468b324ec1031385ce04e4'),
 'amenities': ['Wifi',
  'Essentials',
  'Kitchen',
  'TV',
  'Hangers',
  'Hair dryer',
  'Washer',
  'Iron',
  'Shampoo',
  'Laptop friendly workspace'],
 'host_id': '99997585',
 'address': {'market': 'Rio De Janeiro'}}

In [70]:
# Check the fields
db.listing_reviews.find_one()

{'_id': '58663741',
 'listing_id': '10006546',
 'reviewer_id': '51483096',
 'reviewer_name': 'Cátia',
 'comments': 'A casa da Ana e do Gonçalo foram o local escolhido para a passagem de ano com um grupo de amigos. Fomos super bem recebidos com uma grande simpatia e predisposição a ajudar com qualquer coisa que fosse necessário.\r\nA casa era ainda melhor do que parecia nas fotos, totalmente equipada, com mantas, aquecedor e tudo o que pudessemos precisar.\r\nA localização não podia ser melhor! Não há melhor do que acordar de manhã e ao virar da esquina estar a ribeira do Porto.',
 'date': datetime.datetime(2016, 1, 3, 5, 0)}

In [71]:
from datetime import datetime
# create the document
doc = {
    "listing_id": property_id,
    "reviewer_id": "67084875",
    'reviewer_name': 'Nick',
    "comments": "This property is amazing!",
    "date": datetime.utcnow()
}
# Add a new review to the property
db.listing_reviews.insert_one(doc)

In [72]:
# Check if the review was saved
review = db.listing_reviews.find_one({"listing_id": property_id})
pprint(review)

{'_id': ObjectId('64468b3e4ec1031385ce04e5'),
 'comments': 'This property is amazing!',
 'date': datetime.datetime(2023, 4, 24, 13, 59, 26, 309000),
 'listing_id': ObjectId('64468b324ec1031385ce04e4'),
 'reviewer_id': '67084875',
 'reviewer_name': 'Nick'}


### Question 15
Add a new review metric called x_factor to the new document and give a score of 10. Show that the average across all metrics is correctly calculated for this listing, with the query you previously developed.

In [73]:
# Adding the new review metric to the new document
db.listingsAndReviews_new.update_one({"_id": property_id}, {"$set": {"review_scores.x_factor": 10}})


# Calculating the average using the query developed in Question 10
result = list(db.listingsAndReviews_new.aggregate([
    {"$match": {"_id": property_id}},
    {"$project":{"review_scores_array": {"$objectToArray": "$review_scores"}}},
    {"$unwind": "$review_scores_array"},
    {"$group": {"_id": "$_id", "avg_review_score": {"$avg": "$review_scores_array.v"}}},
    {"$project": {"_id": 0, "avg_review_score": {"$round": ["$avg_review_score", 2]}}}
]))

pprint(result)

[{'avg_review_score': 10.0}]
